File: backend/assembler.py
import re
from pathlib import Path

REPO_ROOT = Path(__file__).parent.parent
BLOCKS_DIR = REPO_ROOT / "cv-ai" / "prompt-blocks"

SECTOR_MAP = [
    (re.compile(r"(soft|develop|engineer|data|devops|\bit\b|yazılım|bilişim|teknoloji)", re.IGNORECASE), "tech"),
    (re.compile(r"(financ|bank|invest|account|audit|finans|muhasebe|denetim)", re.IGNORECASE), "finance"),
    (re.compile(r"(design|market|brand|content|advert|creative|tasarım|pazarlama|reklam)", re.IGNORECASE), "creative"),
    (re.compile(r"(logist|warehouse|manufact|construct|retail|driver|production|lojistik|üretim|inşaat|perakende)", re.IGNORECASE), "operational"),
]

def pick_persona(seniority: str) -> str:
    if seniority == "academic":
        return "academic"
    elif seniority == "executive":
        return "executive"
    elif seniority in ("senior", "mid"):
        return "experienced"
    elif seniority == "entry":
        return "early_career"
    else:
        raise ValueError(f"Unknown seniority: {seniority}")

def pick_sector(target_sector: str) -> str:
    for pattern, sector in SECTOR_MAP:
        if pattern.search(target_sector):
            return sector
    return "general"

def pick_localization(target_region: str, sector: str, company_type_hint: str = "unknown") -> str:
    if company_type_hint == "multinational":
        return "intl_multinational"
    if company_type_hint == "domestic":
        return "tr_domestic"
    
    # Inference fallback: region outside Turkey OR tech sector (multinational/remote-heavy)
    abroad = not bool(re.search(r"türkiye|turkey|tr\b|istanbul|ankara|izmir", target_region, re.IGNORECASE))
    if abroad:
        return "intl_multinational"
    if sector == "tech":
        return "intl_multinational"
    return "tr_domestic"

def select_persona_profile(current_title: str, target_role: str, target_sector: str, target_region: str, seniority: str, company_type_hint: str = "unknown") -> dict:
    persona = pick_persona(seniority)
    sector_overlay = pick_sector(target_sector)
    localization = pick_localization(target_region, sector_overlay, company_type_hint)
    return {
        "persona": persona,
        "sector_overlay": sector_overlay,
        "localization": localization,
        "example": persona,
        "output_language": "tr" if localization == "tr_domestic" else "en"
    }

def load_block(*parts: str) -> str:
    path = BLOCKS_DIR.joinpath(*parts)
    return path.read_text(encoding="utf-8").strip()

def assemble_prompt(profile: dict, mode: str) -> str:
    segments = [
        load_block("core", "core.md"),
        load_block("core", "output-contract.md"),
        load_block("personas", f"{profile['persona']}.md"),
        load_block("sectors", f"{profile['sector_overlay']}.md"),
        load_block("localization", f"{profile['localization']}.md"),
        load_block("examples", f"{profile['example']}.md"),
        f'# RUNTIME MODE\nmode: "{mode}"\noutput_language: "{profile["output_language"]}"'
    ]
    return "\n\n---\n\n".join(segments)

def build_score_call(router_answers: dict) -> tuple:
    profile = select_persona_profile(
        current_title=router_answers.get("current_title", ""),
        target_role=router_answers.get("target_role", ""),
        target_sector=router_answers.get("target_sector", ""),
        target_region=router_answers.get("target_region", ""),
        seniority=router_answers.get("seniority", "mid"),
        company_type_hint=router_answers.get("company_type_hint", "unknown")
    )
    system_prompt = assemble_prompt(profile, "score")
    return profile, system_prompt

def build_rebuild_call(stored_profile: dict) -> str:
    return assemble_prompt(stored_profile, "rebuild")

File: cv-ai/assembler.ts
/**
 * assembler.ts
 * --------------------------------------------------------------------------
 * Deterministic block selection + prompt assembly for the CV AI architecture.
 *
 * Pipeline:
 *   router answers + CV  ->  selectPersonaProfile()  ->  persona_profile
 *   persona_profile      ->  assemblePrompt()        ->  one system prompt
 *
 * The persona_profile is computed ONCE (in score mode), persisted on the
 * session, and reused unchanged in rebuild mode. This guarantees Phase 1 and
 * Phase 2 see the same rubric. Never call selectPersonaProfile() again in
 * rebuild mode — load the stored profile instead.
 * --------------------------------------------------------------------------
 */

import { readFileSync } from "node:fs";
import { join } from "node:path";

const BLOCKS = join(__dirname, "prompt-blocks");

type Seniority = "entry" | "mid" | "senior" | "executive" | "academic";
type Persona = "early_career" | "experienced" | "executive" | "academic";
type Sector = "tech" | "finance" | "creative" | "operational" | "general";
type Localization = "tr_domestic" | "intl_multinational";
type Mode = "score" | "rebuild";

interface RouterAnswers {
  current_title: string;
  target_role: string;
  target_sector: string;
  target_region: string;
  seniority: Seniority;                  // explicit — never model-inferred
  company_type_hint?: "domestic" | "multinational" | "unknown";
}

interface PersonaProfile {
  persona: Persona;
  sector_overlay: Sector;
  localization: Localization;
  example: Persona;                      // example always matches persona
  output_language: "tr" | "en";
}

/* ------------------------------------------------------------------ */
/* 1. SELECTOR — pure, deterministic. Maps answers -> persona_profile. */
/* ------------------------------------------------------------------ */

function pickPersona(seniority: Seniority): Persona {
  switch (seniority) {
    case "academic":   return "academic";
    case "executive":  return "executive";
    case "senior":     return "experienced";   // senior IC/manager -> experienced rubric
    case "mid":        return "experienced";
    case "entry":      return "early_career";
  }
}

// Map free-text sector to one of the five overlays. Keep this table in code,
// not in the prompt — it must be deterministic and testable.
const SECTOR_MAP: Array<[RegExp, Sector]> = [
  [/(soft|develop|engineer|data|devops|it|yazılım|bilişim|teknoloji)/i, "tech"],
  [/(financ|bank|invest|account|audit|finans|muhasebe|denetim)/i,       "finance"],
  [/(design|market|brand|content|advert|creative|tasarım|pazarlama|reklam)/i, "creative"],
  [/(logist|warehouse|manufact|construct|retail|driver|production|lojistik|üretim|inşaat|perakende)/i, "operational"],
];

function pickSector(targetSector: string): Sector {
  for (const [re, sector] of SECTOR_MAP) if (re.test(targetSector)) return sector;
  return "general";
}

// Domestic vs multinational. Explicit hint wins; otherwise infer.
function pickLocalization(a: RouterAnswers, sector: Sector): Localization {
  if (a.company_type_hint === "multinational") return "intl_multinational";
  if (a.company_type_hint === "domestic")      return "tr_domestic";

  // Inference fallback: region outside Turkey OR tech sector (multinational/remote-heavy)
  const abroad = !/türkiye|turkey|tr\b|istanbul|ankara|izmir/i.test(a.target_region);
  if (abroad) return "intl_multinational";
  if (sector === "tech") return "intl_multinational"; // tech skews multinational in TR
  return "tr_domestic";
}

export function selectPersonaProfile(a: RouterAnswers): PersonaProfile {
  const persona = pickPersona(a.seniority);
  const sector_overlay = pickSector(a.target_sector);
  const localization = pickLocalization(a, sector_overlay);
  return {
    persona,
    sector_overlay,
    localization,
    example: persona,                                   // example always matches persona
    output_language: localization === "tr_domestic" ? "tr" : "en",
  };
}

/* ------------------------------------------------------------------ */
/* 2. ASSEMBLER — concatenates the selected blocks into one prompt.    */
/* ------------------------------------------------------------------ */

function load(...parts: string[]): string {
  return readFileSync(join(BLOCKS, ...parts), "utf8").trim();
}

export function assemblePrompt(profile: PersonaProfile, mode: Mode): string {
  // Order matters: core rules first, then the rubric, then modifiers, then the
  // wrapper, then the calibration example. Later blocks refine earlier ones.
  const segments = [
    load("core", "core.md"),
    load("core", "output-contract.md"),
    load("personas", `${profile.persona}.md`),
    load("sectors", `${profile.sector_overlay}.md`),
    load("localization", `${profile.localization}.md`),
    load("examples", `${profile.example}.md`),
    `# RUNTIME MODE\nmode: "${mode}"\noutput_language: "${profile.output_language}"`,
  ];
  return segments.join("\n\n---\n\n");
}

/* ------------------------------------------------------------------ */
/* 3. ENTRY POINTS                                                     */
/* ------------------------------------------------------------------ */

// Phase 1: compute profile, assemble, return BOTH so the profile can be persisted.
export function buildScoreCall(a: RouterAnswers) {
  const profile = selectPersonaProfile(a);
  const systemPrompt = assemblePrompt(profile, "score");
  return { profile, systemPrompt };   // persist `profile` on the session
}

// Phase 2: DO NOT re-select. Load the stored profile and reuse it.
export function buildRebuildCall(storedProfile: PersonaProfile) {
  const systemPrompt = assemblePrompt(storedProfile, "rebuild");
  return { systemPrompt };
}
